autotools.py/autoreconf: Show the depends_on()s to add to the package

bernhardkaindl opened the pull request:
Based on the commit to fix missing autoreconf depends in #26101:
    
Generate a template for cut-and-paste into the package.py to fix it:
```py 
=> numactl: Executing phase: 'autoreconf'
==> [date-time] Warning: *********************************************************
*** Please add these lines to the depends of numactl/package.py:
    depends_on('autoconf', type='build', when='@2.0.14')
    depends_on('automake', type='build', when='@2.0.14')
    depends_on('libtool', type='build', when='@2.0.14')
*** and tweak the version in when='@...' as needed.
==> [date-time] Warning: *********************************************************
==> Error: RuntimeError: Cannot generate configure: missing dependencies ['autoconf', 'automake', 'libtool']
```

bernhardkaindl committed:
build_systems/autotools.py: Fix detection of missing tools for autoreconf

If the `configure` script is not present, the `autoreconf()` phase runs
the `autoreconf` tool, generating the `configure` script and other files.

Because `autoconf`, `automake` and `libtool` are needed to run `autoreconf`,
`AutotoolsPackage.autoreconf()` attempts to check that these are available.

Issue #26005 shows that this check currently does not work properly:
It needs to look for the required dependencies in `specs._dependencies`
instead of currently in 'specs'.

Therefore it currently wrongly assumes that the autotools are added
as dependencies, when in reality, they are not. If the autotools are
then not installed properly by the host system, strange errors like
the error shown in issue #26005 occur. This fixes the dependency check.

haampie commented:
Can you make it part of the exception itself? Then it's shown without `-v`. E.g.

```diff
diff --git a/lib/spack/spack/build_systems/autotools.py b/lib/spack/spack/build_systems/autotools.py
index 8edbfe1de7..531771ea5b 100644
--- a/lib/spack/spack/build_systems/autotools.py
+++ b/lib/spack/spack/build_systems/autotools.py
@@ -252,6 +252,18 @@ def delete_configure_to_force_update(self):
         if self.force_autoreconf:
             force_remove(self.configure_abs_path)
 
+    def _autoreconf_warning(self, spec, missing):
+        msg = ("Cannot generate configure: missing dependencies {0}. Please add "
+               "the following lines to the package:\n\n".format(", ".join(missing)))
+
+        for dep in missing:
+            msg += ("    depends_on('{0}', type='build', when='@{1}')\n"
+                    .format(dep, spec.version))
+
+        msg += "\nand tweak the version in when='@...' as needed."
+
+        return msg
+
     def autoreconf(self, spec, prefix):
         """Not needed usually, configure should be already there"""
         # If configure exists nothing needs to be done
@@ -261,9 +273,10 @@ def autoreconf(self, spec, prefix):
         needed_dependencies = ['autoconf', 'automake', 'libtool']
         build_deps = [d.name for d in spec.dependencies(deptype='build')]
         missing = [x for x in needed_dependencies if x not in build_deps]
+
         if missing:
-            msg = 'Cannot generate configure: missing dependencies {0}'
-            raise RuntimeError(msg.format(missing))
+            raise RuntimeError(self._autoreconf_warning(spec, missing))
+
         tty.msg('Configure script not found: trying to generate it')
         tty.warn('*********************************************************')
         tty.warn('* If the default procedure fails, consider implementing *')
```

Then the output is:

![Screenshot from 2021-09-21 16-50-43](https://user-images.githubusercontent.com/194764/134193899-7c357eb5-3d2d-4cc1-83b1-21e71add2bd8.png)

bernhardkaindl committed:
autotools.py/autoreconf: Show the depends_on()s to add to the package

A previous commit fixed detection of missing tools for autoreconf.

This commit shows a template for cut-and-paste into the package to fix it:

```py
==> fast-global-file-status: Executing phase: 'autoreconf'
==> Error: RuntimeError: Cannot generate configure: missing dependencies autoconf, automake, libtool.

Please add the following lines to the package:

    depends_on('autoconf', type='build', when='@master')
    depends_on('automake', type='build', when='@master')
    depends_on('libtool', type='build', when='@master')

Update the version (when='@master') as needed.
```

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

bernhardkaindl commented:
> Can you make it part of the exception itself? Then it's shown without `-v`. E.g.

This is cool! :+1:  :1st_place_medal: :100:  I:
- moved `"Please add"` to an new line,
- replaced `"and tweak"` with `"Update"` and
- use the spec.version for the `"Update the version"` line too.

New commit message with the new output:

```py
    This commit shows a template for cut-and-paste into the package to fix it:

    ```py
    ==> fast-global-file-status: Executing phase: 'autoreconf'
    ==> Error: RuntimeError: Cannot generate configure: missing dependencies autoconf, automake, libtool.
    
    Please add the following lines to the package:
    
        depends_on('autoconf', type='build', when='@master')
        depends_on('automake', type='build', when='@master')
        depends_on('libtool', type='build', when='@master')
    
    Update the version (when='@master') as needed.
    ```
    
    Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>
```
:+1:

alalazo reviewed (dismissed):
Thanks!

haampie commented:
Thanks, this is a huge quality of life improvement :)

haampie commented:
@spackbot run pipeline

haampie commented:
@spackbot run pipeline

bernhardkaindl commented:
@spackbot run pipeline

haampie commented:
@spackbot run pipeline

haampie commented:
@scottwittenburg is 14 hours for a pipeline normal? :sweat_smile:

scottwittenburg commented:
@haampie No it's not normal, it's due to the E4S mirror index getting further and further out of date, which is a result of issues with splicing old specs.

bernhardkaindl commented:
I saw it has one failed job (a script failure) in the pipeline: https://gitlab.spack.io/spack/spack/-/jobs/997549
```
==> [2021-09-22-08:19:14.935290] find_matching_specs: about to parse specs for /2irmcuao37qogct5eqi65lobebo7a4rs
==> [2021-09-22-08:19:14.936324] Reading config file /builds/spack/spack/etc/spack/defaults/bootstrap.yaml
==> [2021-09-22-08:19:14.945775] DATABASE LOCK TIMEOUT: 3s
==> [2021-09-22-08:19:14.945820] PACKAGE LOCK TIMEOUT: No timeout
==> [2021-09-22-08:19:14.947807] SpackError: No installed spec matches the hash: '2irmcuao37qogct5eqi65lobebo7a4rs'
==> [2021-09-22-08:19:14.947842] Error: No installed spec matches the hash: '2irmcuao37qogct5eqi65lobebo7a4rs'
```
@haampie Should we restart the pipeline?

alalazo commented:
@bernhardkaindl The pipeline is under maintainance for part of today, so I'd give it some more time (sorry). If by tomorrow the situation is not solved we can think of going through PRs and "force" merge the ones that pass all the other tests + won't affect Gitlab pipelines.

scottwittenburg commented:
@bernhardkaindl I see this message in the job trace of the job you linked above:

```
==> [2021-09-22-08:19:07.970297] numactl: Executing phase: 'autoreconf'
==> [2021-09-22-08:19:12.767781] Error: ProcessError: Command exited with status 1:
    '/usr/bin/bash' './autogen.sh'
See build log for details:
  /tmp/root/spack-stage/spack-stage-numactl-2.0.14-2irmcuao37qogct5eqi65lobebo7a4rs/spack-build-out.txt
Traceback (most recent call last):
  File "/builds/spack/spack/lib/spack/spack/build_environment.py", line 1022, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "/builds/spack/spack/lib/spack/spack/installer.py", line 1883, in build_process
    return installer.run()
  File "/builds/spack/spack/lib/spack/spack/installer.py", line 1749, in run
    self._real_install()
  File "/builds/spack/spack/lib/spack/spack/installer.py", line 1847, in _real_install
    phase(pkg.spec, pkg.prefix)
  File "/builds/spack/spack/lib/spack/spack/package.py", line 116, in phase_wrapper
    phase(spec, prefix)
  File "/builds/spack/spack/var/spack/repos/builtin/packages/numactl/package.py", line 38, in autoreconf
  File "/builds/spack/spack/lib/spack/spack/util/executable.py", line 225, in __call__
    raise ProcessError('Command exited with status %d:' %
spack.util.executable.ProcessError: Command exited with status 1:
    '/usr/bin/bash' './autogen.sh'
```

Somehow even though that failed, `spack install` exited non-zero so the CI code continued on afterwards, attempting to create a binary package for software that failed to even configure.  Hence, the error message was actually a long way from the bottom where you likely expected to see it.  Sorry about that 😞 

If you browse the artifacts of the failed job, you can find `jobs_scratch_dir/logs/spack-build-out.txt` which in this case has the following contents:

```
==> [2021-09-22-08:19:07.970297] numactl: Executing phase: 'autoreconf'
==> [2021-09-22-08:19:07.971902] '/usr/bin/bash' './autogen.sh'
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: linking file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: linking file 'm4/libtool.m4'
libtoolize: linking file 'm4/ltoptions.m4'
libtoolize: linking file 'm4/ltsugar.m4'
libtoolize: linking file 'm4/ltversion.m4'
libtoolize: linking file 'm4/lt~obsolete.m4'
Useless use of a variable in void context at /home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu20.04-ppc64le/gcc-9.3.0/automake-1.16.3-rzfsznblqwmptsoua2hoy6rxgqevce5j/bin/automake line 8405.
autoreconf: automake failed with exit status: 255
```

It makes me think there could be a real problem, and that re-running the pipeline is likely fail in the same way.

🤷 Hope it helps.

bernhardkaindl commented:
> @bernhardkaindl I see this message in the job trace of the job you linked above: ... 
> It makes me think there could be a real problem, and that re-running the pipeline is likely fail in the same way.
> 
> 🤷 Hope it helps.

Possibly, but since it's only numactl, and I built numactl (and hundreds of other autoconf packages, and the pipeline ran otherwise fine), and the fact that the code change does not appear to be connected to this fail, I'm not eager to re-run the pipeline.

I'd then consider following @alalazo's comment above, force the commit and if then numactl needs a fix for powerpc64le, that's a different PR. And I assume the fail might have to do with maybe some error in the autoconf/automake install, where there were fetching errors in the log above that location.

haampie commented:
@spackbot run pipeline

haampie commented:
@spackbot run pipeline
